### ISSUE TITLE
tcp zerowindow always with opentsdb backend

### ISSUE THREAD

**Issue by cryptofuture** [firehol/netdata#1432]
I tried different ipv6 tunnels, but ether way always tcp zerowindow with opentsdb backend.  Connection ok with packets without metrics, but packets with metrics always out of the window. I tried play wit rmem/wmem and it makes no difference.  
As far as I know ipv6 takes window settings from ipv4, and I already have tcp windows size tunned with ipv4.  
More above I tried to collect metrics from two different netdata installations, tcp zerowindow ether way. I can see packets with metrics, but seems like I have "You shall not pass" situation.

**Comment by ktsaou** [comment]
Can you explain your setup? Are you giving an IPv6 target as a netdata backend?

**Comment by cryptofuture** [comment]
netdata host1 (ipv6 he.net tunnel)  -> opentsdb backend (ipv6 he.net tunnel)  
netdata host2 (ipv6 he.net tunnel) -> same opentsdb backend (ipv6 he.net tunnel)  
First I thought problem was because I used miredo on backend, but now I have very stable tunnels.   And I also tried another 6in4 provider, but stopped with he.net, never had problems with them. And I have tcp the, zero window with all packets with metrics inside with all tunnels I tried.  
netdata host1 and host2 located in different counties, and backend in different physical location too.  Also I didn't tried to mess with openvpn there, since openvpn over ipv6 tunnel will be even slower.
Btw, I may switch to proper server for opentsdb backend with static ipv4 address, but I need to collect some statics about how much space I'll need and how loaded opentsdb backend will be.  
Hosts with netdata are proper servers one with 1Gbit and another with 10Gbit NICs.  
And yes, I just to forward port is not and option on backend, since its located in DMZ, and port forwarding only works when my VPN is off.
Basically opentsdb backend is just [lxd container](https://github.com/cryptofuture/lxd-image-opentsdb) I run on localhost in current setup.
And yes, I never had tcp zerowindow problems inside ipv6 tunnels before.

**Comment by ktsaou** [comment]
Well, the code I used I pretty generic: https://github.com/firehol/netdata/blob/c72ab23ee2fd1d8af22a5362fe947c54c3c683ef/src/backends.c#L43-L84

I haven't used anything to set a TCP Window size or anything similar.

Is the tunnel working for other TCP connections? I mean, everything else works (like ssh, curl, etc) and netdata fails?

**Comment by cryptofuture** [comment]
curl works for sure, and first 84/92/72 bytes packet works too. I think thats what I see next like client connected in opentsdb log packet. Next packets with metrics are 1480 bytes (max size for henet tunnel)

**Comment by ktsaou** [comment]
I am reading again about it... I found this line: https://github.com/firehol/netdata/blob/c72ab23ee2fd1d8af22a5362fe947c54c3c683ef/src/backends.c#L74

Well, the documentation says:

```
For IPv6 sockets only.  This is the index corresponding to the
network interface on which to exchange datagrams.  This
parameter is ignored for IPv4 sockets or when an IPv6 "scoped
address" is specified in 'host' (i.e. where the colon-hex
network address is augmented with the scope ID).
```

I gave the value `0`, which means "the first network interface". Could you please comment out that line and retry?

**Comment by ktsaou** [comment]
Once you comment it out, please try to specify the target IP address as:

```
hex:hex:...:hex%interface
```

so, append `%interface`.

If you also have a port, use:

```
[hex:hex:...:hex%interface]:PORT
```

But for sure, that line must be commented for the above to work.

**Comment by cryptofuture** [comment]
Will try it definitely, since I need to make it work anyway. Just a little pre-xmas mess didn't live well along with my plans.

**Comment by cryptofuture** [comment]
With `/*name.sin6_scope_id = 0; */` commented I have 
```
Cannot resolve host '/* backend_ipv6_ip%he-ipv6', port '35404': Name or service not known

2016-12-24 12:41:16: netdata: ERROR: Failed to update database backend '[backend_ipv6_ip%he-ipv6]:35404'
```
With `name.sin6_scope_id = 1;`  and `[hex:hex:...:hex]:PORT` netdata host1 works. Testing on netdata host2 now.

**Comment by ktsaou** [comment]
hm... then it is a bug... Let me find a way to fix this...

**Comment by cryptofuture** [comment]
No changes with both host communication, still zerowindow. Looks like first 4 packets with metrics worked, and then TCP DUP ACK, Window Update, ZeroWindow. I could share pcap file with you privately as email attachment for ex.

**Comment by cryptofuture** [comment]
UPD: I just send you tcpdumps on email from GitHub profile, hope you not against it.

**Comment by cryptofuture** [comment]
Found something could be useful on stackoverflow:
>I've seen this happen if you are not emptying the network buffer correctly at the client end and the TCP window size is reduced to zero as a consequence. 
Will try with TCP_NODELAY now.

**Comment by ktsaou** [comment]
could you please checkout `https://github.com/ktsaou/netdata` and try it?
I have re-worked the connection entirely.

The format to specify the interface is now:

```
/ connect_to()
//
// definition format:
//
//    [PROTOCOL:]IP[%INTERFACE][:PORT]
//
// PROTOCOL  = tcp or udp
// IP        = IPv4 or IPv6 IP or hostname, optionaly enclosed in [] (required for IPv6)
// INTERFACE = for IPv6 only, the network interface to use
// PORT      = port number or service name
```

so, you can use: `tcp:[11:22:..:ee:ff]%eth0:12345`

**Comment by ktsaou** [comment]
wait a moment... I have a bug...

**Comment by cryptofuture** [comment]
k, waiting.

**Comment by ktsaou** [comment]
ok, you can test it.
I have verified it works on IPv4 so, please check that it works in your case.

**Comment by cryptofuture** [comment]
Its a little bit better, still a lot of zerowindow, new tcp out of order packets. I also can't figure out why so may packets with `put: No such names for metrics` context.

**Comment by cryptofuture** [comment]
I just had an idea, opentsdb works with TCP_NODELAY by default. 
```
# Disable Nagel's algorithm.  Default is True
tsd.network.tcp_no_delay = true
```
testing with false now, mb it will change something.

**Comment by cryptofuture** [comment]
Its even worse without NODELAY(with `tsd.network.tcp_no_delay = false`) from opentsdb side. Was like 10-15 success rate before, now 0.

**Comment by ktsaou** [comment]
The TCP_NODELAY is about minimizing latencies. I don't think it can influence anything else.

Could you please capture a whole session with tcpdump? That is:

1. stop netdata
2. start capture
3. start netdata
4. wait of it to happen
5. stop capture and send me the pcap

**Comment by ktsaou** [comment]
Also, is netdata complaining at its error.log?

**Comment by cryptofuture** [comment]
Done, check email.
```
2016-12-24 16:35:05: alarm-notify.sh: INFO: not sending notification for CLEAR on 'netdata.backend_metrics.backend_metrics_lost' (last status was UNINITIALIZED)
2016-12-24 16:35:05: netdata: ERROR: child pid 23338 exited with code 1. (errno 9, Bad file descriptor)
2016-12-24 16:35:05: alarm-notify.sh: INFO: not sending notification for CLEAR on 'netdata.backend_metrics.backend_metrics_sent' (last status was UNINITIALIZED)
2016-12-24 16:35:05: netdata: ERROR: child pid 23342 exited with code 1. (errno 9, Bad file descriptor)
```

**Comment by ktsaou** [comment]
I think you need to enable metrics auto-creation for opentsdb. I think opentsdb is doing this, because netdata sends metrics it is not allowed to accept.

Check this: https://github.com/OpenTSDB/opentsdb/issues/792#issuecomment-218681567

**Comment by ktsaou** [comment]
And this: https://groups.google.com/forum/#!topic/opentsdb/cI_nAl0nOVc

**Comment by ktsaou** [comment]
And I think there is another issue. Normally, these databases do not reply back. But in your case, the databases is complaining it is not accepting these metrics. netdata however, does not read these replies, so the buffers get full and I think this results to the zero window size.

I am modifying netdata now to read the replies too...

**Comment by cryptofuture** [comment]
I have some crazy idea, I think I'll try netdata backend mechanism with KairosDB under the same ipv6 tunnels. I enabled metrics autocreation, but that didn't helped much, that just delayed time for zerowindow packets a bit. Looks like KairosDB support graphite and telnet put like opentsdb, and I can connect it to grafana too.

**Comment by cryptofuture** [comment]
Btw, thanks for great help, and light, but powerful and totally awesome netdata. Wish you best holidays and Christmas ever.   
In meantime, please don't close this thread I'll report on KairosDB under the same setup experience there.

**Comment by ktsaou** [comment]
I added the code to dequeue the receive buffer from the socket. I think your opentsdb is chatty, so it fills the receive queue which results to this zero window situation.

Could you please check. Check the netdata error log. It should be logging it is receiving data from the server.

**Comment by ktsaou** [comment]
added new metrics to the backend charts to monitor reads and bytes received.
Also, verified the new code works on both graphite and opentsdb without issues (IPv4 only).

So, it would be nice if you could confirm it works for you on IPv6.

**Comment by cryptofuture** [comment]
![shoot](https://cloud.githubusercontent.com/assets/6484506/21469121/a0231df8-ca32-11e6-9926-777c7d8f8579.png)
Judging by fresh dump without tcp zerowindow packets things look ok. Metrics in netdata shows things really awesome, like 0,3ms latency, but in fact ping6 to backend from netdata host1 120ms and 200ms from host2. Overall from 6.6mb tcpdump I have like 30% success packets (without counting retransmission, out of window, tcp out of order and etc). Mb thats quite ok for ipv6 tunnels, but I blame opentsdb. But well at least I finally can receive metrics! Problem fixed definitely, but I have some expectations on KairosDB. And if my thoughts right, I could test both plaintext and telnet connections to KairosDB and compare network performance with opentsdb data.

**Comment by ktsaou** [comment]
ok nice.

regarding the timings netdata shows, it can only measure the time required to send the data. It does not have any means to know when the backend received them. If the backend server is congested and netdata is sending them at a rate higher than it can consume, then the O/S buffers will start filling up and higher latencies will be reported. Probably the chart is misleading and should be removed...

I see on the charts that there is not any "read" operations and bytes, so opentsdb does not respond anything back. So, I guess my changes did the trick.

Regarding the 30% success rate of your tunnel, my suggestion is to fix it. The only acceptable rate is 100% for any IP version... There is definitely something very wrong there.

Thanks!

**Comment by ktsaou** [comment]
cleaned up the code and removed the latency chart.

Is it ok to close this ticket now?

**Comment by ktsaou** [comment]
PR #1439 is merged now, so you can use the main repo again.
Thanks for reporting this problem.

**Comment by cryptofuture** [comment]
Yes, sure, anyway if I remember right, I could comment in closed issue.

**Comment by cryptofuture** [comment]
After all tests I ended with InfluxDB  with opentsdb plugin (no need extra installation, setup in influxdb config) with grafana (chronograf is still too raw) and communication via openvpn servers on netdata hosts over ipv4. 
So, yes netdata could work with InfluxDB without extra hassle. And as for KairosDB it looks like quite active project, but shown itself quite bad in my setup.  
This is for ex. how netdata's system.cpu looks in grafana:
![cpu](https://cloud.githubusercontent.com/assets/6484506/21546984/f773a48c-cdda-11e6-85af-7643782fe93c.png)  
Mb change https://github.com/firehol/netdata/wiki/netdata-backends page from backends to backend formats, since netdata works fine will all software that support graphite and opentsdb formats. Or mb just name all popular software on wiki page  with graphite/opentsdb formats support. 
For ex.:
```
 1. **graphite** - plaintext interface (Graphite, InfluxDB, KairosDB, Blueflood and other software with graphite support)

      metrics are sent to graphite as `prefix.hostname.chart.dimension`. `prefix` is configured below, `hostname` is the hostname of the machine.

   2. **opentsdb** - telnet interface (OpenTSDB, InfluxDB, KairosDB and other software with opentsdb support)
```
What i mean now it looks like netdata support only OpenTSDB and Graphite, but actually its not true, and netdata support any DBs/system with compatible with opentsdb/graphite format (netdata rocks again).

**Comment by ktsaou** [comment]
Thanks! Nice you reported back.

I changed the wiki according to your suggestions (you could update the wiki too - and I actually welcome you to make changes).

Thanks!

**Comment by cryptofuture** [comment]
I made update on [backends wiki page](https://github.com/firehol/netdata/wiki/netdata-backends)
Finally was able to write some [blog post](https://blog.hda.me/2017/01/09/using-netdata-with-influxdb-backend.html) about it
And yes, grafana dashboards for netdata: (hope links works, didn't tested without cookies yet)
[netdata](https://grafana.net/dashboards/1295)
[lxd container](https://grafana.net/dashboards/1291)  
Feel free to reuse part of post if you want in wiki. I think will be too huge for wiki, if to copypaste it directly, so I just added a link with description.
